File: hat/assets/js/apps/Iaso/domains/stock/types/stocks.ts
type Project = {
    id: number;
    name: string;
};
type OrgUnitType = {
    id: number;
    name: string;
};
type Form = {
    id: number;
    name: string;
};
type StockKeepingUnitChild = {
    id: number;
    name: string;
    value: number;
};
type User = {
    id: number;
    name: string;
};

export type StockKeepingUnit = {
    id: number;
    name: string;
    short_name: string;
    projects: Project[];
    org_unit_types: OrgUnitType[];
    forms: Form[];
    children: StockKeepingUnitChild[];
    display_unit?: string;
    display_precision: number;
    created_at: number;
    created_by: User;
    updated_at: number;
    updated_by: User;
};

export type StockKeepingUnitDto = {
    id: number;
    name: string;
    short_name: string;
    projects: number[];
    org_unit_types: number[];
    forms: number[];
    children: number[];
    display_unit?: string;
    display_precision: number;
};

export type StockItemRuleDto = {
    id: number;
    sku: number;
    form: number;
    question: string;
    impact: Impact;
    order: number;
};

type ShortSKU = {
    id: number;
    name: string;
    short_name: string;
};

export type Impact = 'ADD' | 'SUBTRACT' | 'RESET';
export type StockItemRule = {
    id: number;
    version_id: number;
    sku: ShortSKU;
    form: Form;
    question: string;
    impact: Impact;
    order: number;
    created_at: number;
    created_by?: User;
    updated_at: number;
    updated_by?: User;
};

export type Status = 'DRAFT' | 'UNPUBLISHED' | 'PUBLISHED';
export type StockRulesVersion = {
    id: number;
    name: string;
    status: Status;
    rules?: StockItemRule[];
    created_at: number;
    created_by?: User;
    updated_at: number;
    updated_by?: User;
    deleted_at?: number;
};

File: iaso/admin/stocks.py
from django.contrib.gis import admin

from iaso.models import (
    StockItem,
    StockItemRule,
    StockKeepingUnit,
    StockKeepingUnitChildren,
    StockLedgerItem,
    StockRulesVersion,
)


@admin.register(StockItem)
class StockItemAdmin(admin.ModelAdmin):
    fields = ("org_unit", "sku", "value", "created_at", "updated_at")
    readonly_fields = ("org_unit", "sku", "value", "created_at", "updated_at")
    list_display = ("org_unit", "sku", "value")
    list_filter = ["sku"]

    def has_add_permission(self, request, obj=None):
        return False


@admin.register(StockItemRule)
class StockItemRuleAdmin(admin.ModelAdmin):
    fields = ("sku", "form", "version", "impact", "question", "created_at", "updated_at", "created_by", "updated_by")
    readonly_fields = ("created_at", "updated_at", "created_by", "updated_by")
    list_display = ("sku", "form", "question", "impact", "version", "created_at")
    list_filter = ("sku", "form", "impact")


@admin.register(StockKeepingUnit)
class StockKeepingUnitAdmin(admin.ModelAdmin):
    fields = (
        "account",
        "name",
        "short_name",
        "projects",
        "org_unit_types",
        "forms",
        "display_unit",
        "display_precision",
        "created_at",
        "updated_at",
        "created_by",
        "updated_by",
        "deleted_at",
    )
    readonly_fields = ("created_at", "updated_at", "created_by", "updated_by")
    list_display = ("name", "short_name", "account")
    list_filter = ("account", "name", "short_name")


@admin.register(StockKeepingUnitChildren)
class StockKeepingUnitChildrenAdmin(admin.ModelAdmin):
    fields = ("parent", "child", "created_at", "updated_at", "created_by", "updated_by")
    readonly_fields = ("created_at", "updated_at", "created_by", "updated_by")
    list_display = ("parent", "child", "value")
    list_filter = ("parent", "child")


@admin.register(StockLedgerItem)
class StockLedgerItemAdmin(admin.ModelAdmin):
    fields = ("rule", "sku", "org_unit", "submission", "question", "impact", "value", "created_at", "created_by")
    readonly_fields = (
        "rule",
        "sku",
        "org_unit",
        "submission",
        "question",
        "impact",
        "value",
        "created_at",
        "created_by",
    )
    list_display = ("rule", "sku", "org_unit", "question", "impact", "value", "created_at")
    list_filter = ("sku", "impact", "rule")

    def has_add_permission(self, request, obj=None):
        return False


@admin.register(StockRulesVersion)
class StockRuleVersionAdmin(admin.ModelAdmin):
    fields = ("account", "name", "status", "created_at", "updated_at", "created_by", "updated_by", "deleted_at")
    readonly_fields = ("created_at", "updated_at", "created_by", "updated_by")
    list_display = ("account", "name", "status")
    list_filter = ("account", "status")
